### ISSUE TITLE
F# 6.0 built in tasks expression internal type inference would make more sense syntactically if it matched up with Taskbuilder.fs

### ISSUE THREAD

**Issue by jbtule** [fsharp/fslang-suggestions#1103]
# F# 6.0 built in tasks type inference doesn't match up with Taskbuilder.fs

I honestly don't know if this is a bug or limitation of the new tasks.  

But with Taskbuilder.fs the following ASP.NET MVC Controller action could be written like so. 

```fsharp
    member this.UserInfo (id: string) : Task<IActionResult> =
        task {
            let! emp = this.AuthenticatedEmployeeForEmail(id)
            match emp with
            | None -> return upcast this.Unauthorized()
            | Some(emp') -> 
                let! extra = employeeRepo.GetExtraVPNByEmail(emp'.EMAIL)
                return upcast this.View({Emp = emp'; VPNs = extra; Conns = Seq.empty})
        }
```

However, switching to the new F# 6.0 tasks, I have to specify the type on almost every upcast (`:>`) instead of once at the method declaration for the compiler to accept it. Workable, but annoying.

```fsharp
    member this.UserInfo (id: string)  =
        task {
            let! emp = this.AuthenticatedEmployeeForEmail(id)
            match emp with
            | None -> return this.Unauthorized() :> IActionResult
            | Some(emp') -> 
                let! extra = employeeRepo.GetExtraVPNByEmail(emp'.EMAIL)
                return this.View({Emp = emp'; VPNs = extra; Conns = Seq.empty})  :> IActionResult
        }
```
## Pros and Cons 

Pro: existing behavior, less repeated type names, more intuitive.

Cons: unsure

## Extra information

Estimated cost (XS, S, M, L, XL, XXL):  ?

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [ ] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by BentTranberg** [comment]
There seems to be nothing here that suggests a specific change to the language or core library, other than "want it to work as I'm used to". I think it would have been better to bring something like this up in one of the F# Slack channels, or possibly on SO, to clarify why there are differences, or whether there's a bug or limitation, or whether a language or library change will possibly make sense, and not least what the details of that change should be.

**Comment by jbtule** [comment]
@BentTranberg I'm leaning towards bug, because Computation Expressions and Resumable State machines usage are syntactically identical, so it seems that type inference should work the same. If this is a limitation, I think it should be in this repo because that is confusing and should be fixed someday.

Feel free to talk about this on the Slack channel, I'm not there, but this really isn't an SO question.

**Comment by BentTranberg** [comment]
For bugs there's the dotnet/fsharp repo. But don't get me wrong, I'm not trying to tell you what to do.

After I went from F# 5 to 6 (and .NET 5 to 6) I've had to make slight changes to source here and there, and it was all about type inference changing the rules slightly, and requiring explicit types. I wonder if this could be the same or a similar issue. Of course if you test it side-by-side and it's still as you've explained, you also disprove this guess of mine.
